src/compiler/checker.ts

```diff
@@ -8223,6 +8459,60 @@ module ts {
             }
         }
 
+        function getPromiseConstructor(node: SignatureDeclaration): EntityName {
+            if (node.type && node.type.kind === SyntaxKind.TypeReference) {
+                var typeReference = <TypeReferenceNode>node.type;
+                return typeReference.typeName;
+            }
+
+            if (globalPromiseSymbol && globalPromiseSymbol === resolveName(node, "Promise", SymbolFlags.Value, undefined, undefined)) {

```

Comment:
Make sure to resolveName *inside* the body of the function, in case there is something inside called Promise.